Add guide on transferring tokens between parachains (#973)
polkadot-developers/polkadot-docs: develop/toolkit/interoperability/paraspell-xcm-sdk/transfer-tokens-between-parachains.md

--8<-- 'code/develop/toolkit/interoperability/paraspell/dry-run-output.html'

## Verify the Existential Deposit

Check if the recipient account meets the [Existential Deposit (ED)](/polkadot-protocol/glossary/#existential-deposit){target=\_blank} requirement before sending by using [`verifyEdOnDestination`](https://paraspell.github.io/docs/sdk/xcmUtils.html#verify-ed-on-destination){target=\_blank}:

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:70:85'
```
Execute the code by running:

```bash
bun run index.ts
```

After that, you will get output confirming the ED:

--8<-- 'code/develop/toolkit/interoperability/paraspell/ed-verification-output.html'

## Get Transfer Info and Fee Estimates

Before sending an XCM transaction, it is helpful to estimate the fees associated with executing and delivering the cross-chain message.

ParaSpell has a helpful function for this: [`getTransferInfo()`](https://paraspell.github.io/docs/sdk/xcmUtils.html#xcm-transfer-info){target=\_blank}. This function returns an estimate of the associated XCM fees, along with the account's balance before and after the fees are paid.

```ts title="index.ts"

> --8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:87:102'

Author: This works for me with an account I create. Try using a custom account and see if that works.

---
title: Transfer Tokens Between Parachains

> description: A step-by-step guide to building, verifying, and executing a transfer from one Parachain to another Parachain using the ParaSpell XCM SDK.

Reviewer: ```suggestion
description: A step-by-step guide to using the ParaSpell XCM SDK to build, verify, and execute a transfer from one Parachain to another.
```

---

# Transfer Tokens Between Parachains

## Introduction

> This guide walks you through transferring tokens between two parachains using the [ParaSpell XCM SDK](https://paraspell.github.io/docs/){target=\_blank}. This example uses [Asset Hub](/polkadot-protocol/architecture/system-chains/asset-hub/){target=\_blank} and the [People Chain](/polkadot-protocol/architecture/system-chains/people/){target=\_blank}, but the same approach can apply to transfers between other parachains.

Reviewer: ```suggestion
This guide walks you through transferring tokens between two parachains using the [ParaSpell XCM SDK](https://paraspell.github.io/docs/){target=\_blank}. This example utilizes [Asset Hub](/polkadot-protocol/architecture/system-chains/asset-hub/){target=\_blank} and the [People Chain](/polkadot-protocol/architecture/system-chains/people/){target=\_blank}. However, the same approach can be applied to transfers between other parachains.
```

For development purposes, this guide will use the [Polkadot TestNet](/develop/networks/#paseo){target=\_blank}, so the transferred token will be PAS.

In this guide, you’ll learn how to:

- Build an XCM transfer transaction using ParaSpell XCM SDK.
- Perform a dry run to validate the transfer.
- Verify the [Existential Deposit (ED)](/polkadot-protocol/glossary/#existential-deposit){target=\_blank} requirement on the destination chain.
- Retrieve information regarding the transfer, along with fee estimates.
- Submit the transaction.

### Prerequisites

- Knowledge of the [fundamentals of Polkadot](/polkadot-protocol/parachain-basics/){target=\_blank}
- Basic understanding of [XCM](/develop/interoperability/intro-to-xcm/){target=\_blank}
- Basic familiarity with JavaScript/TypeScript
- Install [bun](https://bun.com/docs/installation){target=\_blank}, a JavaScript/TypeScript package manager

## Initialize Your Project

Create the project folder:

```bash
mkdir paraspell-transfer
cd paraspell-transfer
```

Initialize the JavaScript project:

```bash
bun init -y
```

Install the required dependencies:

```bash

> bun add @paraspell/sdk@11.3.2 polkadot-api@1.17.1 @polkadot-labs/hdkd-helpers@0.0.25 @polkadot-labs/hdkd@0.0.24

Reviewer: ```suggestion
bun add @paraspell/sdk@11.3.2 polkadot-api@1.17.1 @polkadot-labs/hdkd-helpers@0.0.25 @polkadot-labs/hdkd@0.0.24
```
@brunopgalvao please open a ticket to see about adding these to variables.yml and using placeholders so these versions are easier to keep up to date. Not a blocker to ship this now though.

```

Now add the following setup code to `index.ts`:

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:1:27'
```

Replace the `INSERT_YOUR_SEED_PHRASE` with the seed phrase from your Polkadot development account.

Be sure to fund this account with some PAS tokens on Paseo's Asset Hub using the [Polkadot Faucet](https://faucet.polkadot.io/?parachain=1000){target=\_blank}.

> !!!note "Security Warning"

Reviewer: ```suggestion
!!!warning "Security Warning"
```

>     Never commit your mnemonic phrase in production code. Use environment variables or secure key management systems.

Reviewer: ```suggestion
    Never commit your mnemonic phrase to production code. Use environment variables or secure key management systems.
```

## Build a Token Transfer Transaction

The next step is to build the transaction that you intend to execute.

In this example, you will transfer 10 PAS tokens from Paseo's Asset Hub to Paseo's People Chain system parachain.

Add the ParaSpell transaction code to your `index.ts` file:

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:29:47'
```

Do not execute it just yet. You will perform a dry run of this transaction first to ensure it works as expected.

## Perform a Dry Run

Dry runs simulate the transaction without broadcasting it, allowing you to confirm success in advance.

Add the following dry run code to your `index.ts` script:

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:49:70'
```

> Go ahead and run the script.

Reviewer: ```suggestion
Run the script using the following command:
```

```bash
bun run index.ts
```

> The result of the dry run will be similar to this:

Reviewer: ```suggestion
The result of the dry run will look similar to the following example output:
```

--8<-- 'code/develop/toolkit/interoperability/paraspell/dry-run-output.html'

## Verify the Existential Deposit

Check if the recipient account meets the [Existential Deposit (ED)](/polkadot-protocol/glossary/#existential-deposit){target=\_blank} requirement before sending by using [`verifyEdOnDestination`](https://paraspell.github.io/docs/sdk/xcmUtils.html#verify-ed-on-destination){target=\_blank}:

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:72:88'
```

> Comment out the `dryRunTransfer()` and execute the `verifyED()` by running:

Reviewer: ```suggestion
Comment out the `dryRunTransfer()` function so that it is not executed again. Then, execute the `verifyED()` by running the following command:
```

```bash
bun run index.ts
```

> After that, you will get output confirming the ED:

Reviewer: ```suggestion
After that, you will get output confirming the ED which will look similar to the following:
```

--8<-- 'code/develop/toolkit/interoperability/paraspell/ed-verification-output.html'

## Get Transfer Info and Fee Estimates

Before sending an XCM transaction, it is helpful to estimate the fees associated with executing and delivering the cross-chain message.

ParaSpell has a helpful function for this: [`getTransferInfo()`](https://paraspell.github.io/docs/sdk/xcmUtils.html#xcm-transfer-info){target=\_blank}. This function returns an estimate of the associated XCM fees, along with the account's balance before and after the fees are paid.

```ts title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:90:106'
```

> Comment out the `verifyED()` and execute the `XcmTransferInfo()` by running:

Reviewer: ```suggestion
Comment out the `verifyED()` function so it doesn't execute again. Then, execute the `XcmTransferInfo()` function by running the following command:
```

```bash
bun run index.ts
```

> You should be able to see all the information for your transfer:

Reviewer: ```suggestion
You will see all the information for your transfer similar to the following example:
```

--8<-- 'code/develop/toolkit/interoperability/paraspell/transfer-info-output.html'

Now that you have:

> - Completed a successful dry run of the transaction

Reviewer: ```suggestion
- Completed a successful dry run of the transaction.
```

> - Verified the existential deposit on the recipient account

Reviewer: ```suggestion
- Verified the existential deposit on the recipient account.
```

> - Obtained an estimate of the associated XCM fees

Reviewer: ```suggestion
- Obtained an estimate of the associated XCM fees.
```

> Now you can execute the transfer function by adding the following statement:

Reviewer: Is something missing here? Or are we pointing at the same item twice and one of these should be deleted?

Author: Fixed.

Add the following code:

```typescript title="index.ts"
--8<-- 'code/develop/toolkit/interoperability/paraspell/index.ts:108:108'
```

> Comment out the `XcmTransferInfo()` and execute the transfer:

Reviewer: ```suggestion
Comment out the `XcmTransferInfo()` function so it doesn't execute again. Then, execute the transfer by running the following command: 
```

```bash
bun run index.ts
```

Your `transfer` function will submit the transaction, and you will get the following output:

--8<-- 'code/develop/toolkit/interoperability/paraspell/transfer-output.html'

> Once the transaction is successfully included in a block, you will see the recipient's account balance updated, and you will receive output similar to the one below.

Reviewer: ```suggestion
Once the transaction is successfully included in a block, you will see the recipient's account balance updated, and you will receive output similar to the following.
```
Words like "above" and "below' are less friendly to screen reader devices. "Previous" and "following" are preferred
